doc/design/resource-dependency.md

```diff
@@ -0,0 +1,343 @@
+
+A service binding trait needs to be applied before creating the component.
+Moreover, this kind of trait ordering is quite general and applies to many other traits we have seen.
+
+### 4. NSQ deployment
+
+An [NSQ](https://github.com/nsqio/nsq) cluster is composed with three components: nsqd, nsqlookup and nsqadmin.
+To set up a `nsq` cluster, one needs to first make sure that the `nsqd` component is up, then start and make sure that the `nsqlookup` component is up, and run the `nsqadmin` component at last.
+nsqlookup consumes the result of nsqd component instance, and nsqadmin consumes the result of nsqlookup component instance.
+If we couldn't describe the dependencies order between the components, the component couldn't run properly.
+
+This is similar to [kubernetes#65502](https://github.com/kubernetes/kubernetes/issues/65502) , the issue describes dependencies between containers on the same Pod, while we are discussing dependencies between components.
+
+
+## Proposal
+
+The overall idea is to have each resource object specify data inputs coming from other resources' fields.
+The OAM runtime will build the dependency graph accordingly and only create resources once the data inputs are ready.
+To achieve this, we propose:
+
+1. Add the following `DataInputs` into each resource.
+
+    ```go
+    type DataInputs []DataInput
+
+    type DataInput struct {
+      // FromDataOutput specifies the name of the DataOutput object in the same namespace.
+      FromDataOutput string `json:"fromDataOutput"`
+
+      // ToFieldPaths specifies an array of fields within this resource object
+      // that will be overwritten by the value of given DataOutput. The type of the
+      // parameter (e.g. int, string) is inferred from the type of these fields;
+      // All fields must be of the same type. Fields are specified as JSON field
+      // paths without a leading dot, for example 'spec.replicas'.
+      // If empty, it means this input is not used at all.
+      // This is allowed for resource ordering purpose.
+      ToFieldPaths []string `json:"toFieldPaths"`
+    }
+    ```
+
+    For example, a web application may specify input of mysql secret in yaml style:
+
+    ```yaml
+    datainputs:
+    - toFieldPaths: ["spec.connSecret"]
+      fromDataOutput: mysql-conn-secret # check below DataOutput of the same name
+    ```
+
+   The above DataInputs will be json-marshaled and put into an annotation key `core.oam.dev/datainputs` of my-web workload:
+
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+      annotations:
+        "core.oam.dev/datainputs": "[{\"toFieldPaths\": [\"spec.connSecret\"],\"fromDataOutput\": \"mysql-conn-secret\"}]"
+    ```
+
+    Note that my-web won't be created since it has a DataInputs dependency to satisfy.
+
+2. Add a new CRD `DataOutput` with the following definition:
```

Comment:
I think this CRD need to be named as `DataInput`. Because it is bind to components who need to consume these data.

Reply:
The name DataInput is already taken in the input resource side. To correspond to that, this is named to DataOutput. I don't think we should change that.